File: golang/day_02/day_02.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

type Reveal struct {
	count int
	color string
}

type Game struct {
	gameId  int
	reveals []Reveal
}

func readFile(filename string) []string {
	actualFilePath := filepath.Join("..", "..", "inputs", "day_02", filename)
	dat, err := os.ReadFile(actualFilePath)

	if err != nil {
		panic(err)
	}

	stringDat := string(dat)

	stringDat = strings.ReplaceAll(stringDat, "\r\n", "\n")

	return strings.Split(stringDat, "\n")
}

func processLine(line string) Game {
	splitLine := strings.SplitN(line, ":", 2)

	gameId := splitLine[0]
	result := splitLine[1]

	gameId = strings.SplitN(gameId, " ", 2)[1]
	gameId = strings.TrimSpace(gameId)
	gameIdInt, err := strconv.Atoi(gameId)
	if err != nil {
		panic(err)
	}

	revealSplit := strings.Split(result, ";")

	reveals := make([]Reveal, len(revealSplit))

	for _, reveal := range revealSplit {
		reveal = strings.TrimSpace(reveal)
		if reveal == "" {
			continue
		}
		revealSplit := strings.Split(reveal, ", ")

		for _, revealPart := range revealSplit {
			revealPartSplit := strings.Split(revealPart, " ")
			countStr := revealPartSplit[0]
			count, err := strconv.Atoi(countStr)
			if err != nil {
				panic(err)
			}
			color := revealPartSplit[1]
			reveals = append(reveals, Reveal{count: count, color: color})
		}
	}

	return Game{gameId: gameIdInt, reveals: reveals}
}

func validateGame(game Game) bool {
	reveals := game.reveals
	for _, reveal := range reveals {
		if reveal.color == "red" && reveal.count > 12 {
			return false
		}
		if reveal.color == "green" && reveal.count > 13 {
			return false
		}
		if reveal.color == "blue" && reveal.count > 14 {
			return false
		}
	}
	return true
}

func partOne() {
	readData := readFile("input.txt")

	var total int

	for _, line := range readData {
		if line == "" {
			continue
		}
		current := processLine(line)

		if validateGame(current) {
			total += current.gameId
		}

	}

	fmt.Println(total)

}

func partTwo() {
	readData := readFile("input.txt")

	var total int

	for _, line := range readData {
		if line == "" {
			continue
		}

		colorCountMap := make(map[string]int)

		colorCountMap["red"] = 0
		colorCountMap["green"] = 0
		colorCountMap["blue"] = 0

		var current Game
		current = processLine(line)

		for _, reveal := range current.reveals {
			color := reveal.color
			count := reveal.count

			if count > colorCountMap[color] {
				colorCountMap[color] = count
			}
		}

		total += colorCountMap["red"] * colorCountMap["green"] * colorCountMap["blue"]

	}

	fmt.Println(total)

}

func main() {
	partOne()
	partTwo()
}

File: elixir/day_02/day_02.ex
defmodule Day02 do
  defp read_input(input) do
    Path.join(["..", "..", "inputs", "day_02", input])
    |> File.read!()
    |> String.replace("\r\n", "\n")
    |> String.split("\n", trim: true)
  end

  defp validate_reveal(count, color) do
    case color do
      "red" -> count <= 12
      "green" -> count <= 13
      "blue" -> count <= 14
      _ -> false
    end
  end

  defp parse_reveal(reveal) do
    reveal
    |> String.split(",", trim: true)
    |> Enum.map(&String.trim/1)
    |> Enum.map(&String.split(&1, " ", trim: true))
    |> Enum.map(fn [count, color] ->
      count = String.to_integer(count)
      {count, color}
    end)
  end

  defp parse_game(game) do
    game
    |> String.split(";", trim: true)
    |> Enum.map(&String.trim/1)
    |> Enum.map(&parse_reveal/1)
  end

  defp process_line(line) do
    [game_id, results] =
      line
      |> String.split(":", trim: true)
      |> Enum.map(&String.trim/1)

    game_id =
      game_id
      |> String.split(" ", trim: true)
      |> Enum.at(1)
      |> String.to_integer()

    game = results |> parse_game()

    [game_id, game]
  end

  def part_one do
    read_input("input.txt")
    |> Enum.map(&process_line/1)
    |> Enum.map(fn [game_id, game] ->
      [
        game_id,
        Enum.all?(game, fn reveals ->
          reveals
          |> Enum.map(fn {count, color} -> validate_reveal(count, color) end)
          |> Enum.all?()
        end)
      ]
    end)
    |> Enum.filter(fn [_, results] -> results end)
    |> Enum.map(fn [game_id, _] -> game_id end)
    |> Enum.sum()
  end

  def part_two do
    read_input("input.txt")
    |> Enum.map(&process_line/1)
    |> Enum.map(fn [_, game] ->
      game
      |> Enum.reduce(
        %{
          "red" => 0,
          "green" => 0,
          "blue" => 0
        },
        fn reveals, acc ->
          reveals
          |> Enum.reduce(acc, fn {count, color}, acc ->
            Map.update(acc, color, count, &max(&1, count))
          end)
        end
      )
      |> Map.values()
      |> Enum.product()
    end)
    |> Enum.sum()
  end
end

Day02.part_one() |> IO.inspect()
Day02.part_two() |> IO.inspect()
